### PR TITLE
Load data from embedded database.

### PR DESCRIPTION
**Pull request by jeschkies**
Summary:
This patch introduces sled's embedded database as a storage option for gurk.

Outstanding
- [x] Test with save and load roundtrip
- ~[ ] Join channels and messages~
- [x] Persist all data

### CODE REVIEW

File: src/storage.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+            let (channel_id, content): (_, LayoutVerified<U64<BigEndian>, &[u8]>) =
+                LayoutVerified::new_from_suffix(&*bytes).unwrap();
+            let content = str::from_utf8(&bytes)?;
+            let message = Message {
+                from: "not saved".to_owned(),
+                message: Some(content.to_owned()),
+                attachments: Vec::new(),
+                arrived_at: Utc::now(),
+            };
```
Comment (jeschkies) [inline src/storage.rs:77]: I started following the example https://github.com/spacejam/sled/blob/master/examples/structured.rs. However, I'm not sure how to extract the channel Id followed by the string.

**Review by jeschkies — COMMENTED** [review]
I'm not sure what benefit we get from sled since bincode is serializing all messages and there is no join. We could just serialize `AppData` with bincode. What do you think?

**Comment by cleeyv** [comment]
Whichever data store is used, ideally there would be a good option available for how to eventually encrypt it on disk (as per https://github.com/boxdot/gurk-rs/issues/6). I checked about this for sled, and relatively recently the main developer there clarified that this won't likely be a core feature of sled: https://github.com/spacejam/sled/issues/1196#issuecomment-731459241 The good news though is that there is a [bincode_aes](https://lib.rs/crates/bincode_aes) crate that could be an option for encrypting the data store if it were in bincode as you suggest. The functionality of bincode_aes seems pretty basic - all they key management would need to be implemented around it, for example - but it's much better than nothing.

**Comment by boxdot** [comment]
I am still not entirely decided whether we want to serialize the data in one shot (like it is done now, just save everything to disk; same approach is done is in this PR) or if we want, to serialize entities independently. The latter approach is more error prone but also very flexible.

@cleeyv About encryption. Thank you for providing the links. I think we will go with a thin wrapper around `sled` like it is done in `presage`.

**Comment by jeschkies** [comment]
>  The latter approach is more error prone but also very flexible.

What kind of flexibility do you have in mind?

**Comment by boxdot** [comment]
> > The latter approach is more error prone but also very flexible.
> 
> What kind of flexibility do you have in mind?

I was thinking about writing to the store only when a single channel/message/other entity changes. And then writing only this entity, instead of serializing the whole data store.

**Comment by boxdot** [comment]
Since this PR is based on the code too far from the main branch, I believe it easier to redo the work and open a new PR. Closing for now.

Thank you for your work. I am pretty sure we can use the ideas from here.

**Pull request closed without merging**